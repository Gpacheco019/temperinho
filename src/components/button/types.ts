import { ButtonHTMLAttributes, MouseEvent } from 'react';

import { ColorVariation } from '../../types';

export type ButtonTypes = ButtonHTMLAttributes<HTMLButtonElement>;

export type ButtonProps = ButtonTypes & {
  customLayoutColor?: string;
  size?: 'big' | 'small' | 'mini';
  variant?: 'primary' | 'secondary' | 'third';
  colorVariation?: ColorVariation;
  iconPosition?: 'left' | 'right';
  fullWidth?: boolean;
  loading?: boolean;
  square?: boolean;
  disabled?: boolean;
  circle?: boolean;
  icon?: JSX.Element;
  timeout?: number;
  onClick?: (event?: MouseEvent<HTMLButtonElement>) => void;
};

export type IconProps = Record<
  | 'primary.orange'
  | 'primary.green'
  | 'secondary.lightGreen'
  | 'secondary.red'
  | 'secondary.darkBlue'
  | 'secondary.blue'
  | 'neutral.black'
  | 'neutral.grey'
  | 'neutral.darkGrey'
  | 'neutral.lightGrey'
  | 'neutral.extraLightGrey'
  | 'neutral.white',
  ButtonProps
>;

export type WrapperProps = {
  hasIcon: boolean;
} & Pick<
  ButtonProps,
  | 'customLayoutColor'
  | 'size'
  | 'fullWidth'
  | 'square'
  | 'variant'
  | 'circle'
  | 'colorVariation'
>;
