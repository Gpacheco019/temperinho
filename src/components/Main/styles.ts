import styled from 'styled-components';

export const Container = styled.div`
  display: flex;

  width: 100%;
  height: 100vh;
`;

export const Content = styled.div`
  width: 100%;
  max-width: ${({ theme }) => theme.spacings.xlarge};

  display: flex;
  flex-direction: column;
  align-items: center;
  justify-content: center;
`;
