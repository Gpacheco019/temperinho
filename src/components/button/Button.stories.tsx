import React from 'react';
import { MdAdd } from 'react-icons/md';

import { useArgs } from '@storybook/addons';
import { StoryFn, Meta } from '@storybook/react';

import { Button } from '.';
import { ButtonProps } from './types';

const meta: Meta = {
  title: 'Components/Button',
  component: Button,
  args: {
    square: false,
    circle: false
  },
  argTypes: {
    customLayoutColor: {
      control: {
        type: 'color'
      }
    },
    variant: {
      control: {
        type: 'radio',
        options: ['primary', 'secondary', 'third']
      }
    },
    size: {
      control: {
        type: 'radio',
        options: ['mini', 'small', 'big']
      }
    },
    colorVariation: {
      control: {
        type: 'radio',
        options: [
          'orange',
          'green',
          'lightGreen',
          'red',
          'darkBlue',
          'blue',
          'black',
          'darkGrey',
          'grey',
          'lightGrey',
          'extraLightGrey',
          'white'
        ]
      }
    },
    onClick: { action: 'onClick called' }
  }
};

export default meta;

export const Template: StoryFn<ButtonProps> = ({ ...args }) => (
  <Button variant="primary" {...args}>
    children
  </Button>
);

export const Square: StoryFn<ButtonProps> = ({ ...args }) => (
  <Button variant="primary" square {...args}>
    <MdAdd size={24} />
  </Button>
);

export const WithIcon: StoryFn<ButtonProps> = (args) => <Button {...args} />;

WithIcon.args = {
  children: 'with icon',
  icon: <MdAdd size={24} />
};

export const Circle: StoryFn<ButtonProps> = (args) => <Button {...args} />;

Circle.args = {
  circle: true,
  square: true,
  icon: <MdAdd size={24} />,
  iconPosition: 'right'
};

export const FullWidth: StoryFn<ButtonProps> = ({ ...args }) => (
  <Button variant="primary" fullWidth {...args}>
    Button
  </Button>
);

export const Disabled: StoryFn<ButtonProps> = ({ ...args }) => (
  <Button variant="primary" disabled {...args}>
    Disabled
  </Button>
);

export const LoadingOnClick: StoryFn<ButtonProps> = ({ ...args }) => {
  const [, updateArgs] = useArgs();

  function handleButtonClick() {
    updateArgs({ loading: true });
  }

  return (
    <Button onClick={handleButtonClick} fullWidth variant="primary" {...args}>
      Loading On Click
    </Button>
  );
};

export const Timeout: StoryFn<ButtonProps> = ({ ...args }) => (
  <Button timeout={10} {...args}>
    children
  </Button>
);
