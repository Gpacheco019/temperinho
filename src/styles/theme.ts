export const theme = {
  boxShadow: {
    modal: '1px 3px 6px rgba(0, 0, 0, 0.1)',
    card: '0px 4px 8px rgba(0, 0, 0, 0.04)'
  },
  borderRadius: {
    button: '0.25rem',
    input: '0.4rem',
    card: '0.8rem'
  },
  border: {
    solid: (color: string) => `0.1rem solid ${color}`,
    dashed: (color: string) => `0.1rem dashed ${color}`
  },
  font: {
    family:
      "Montserrat, -apple-system, BlinkMacSystemFont, 'Segoe UI', Roboto, Oxygen, Ubuntu, Cantarell, 'Open Sans', 'Helvetica Neue', sans-serif",
    regular: 400,
    medium: 500,
    semibold: 600,
    bold: 700,
    bolder: 900,
    sizes: {
      xxsmall: '1rem',
      xsmall: '1.2rem',
      small: '1.4rem',
      medium: '1.6rem',
      large: '1.8rem',
      xlarge: '2.0rem',
      xxlarge: '2.8rem',
      huge: '5.2rem'
    }
  },
  colors: {
    primary: {
      orange: '#f58031'
    },
    secondary: {
      green: '#008014',
      lightGreen: '#94d376',
      red: '#e74c3c',
      darkBlue: '#1d2236',
      blue: '#3498db'
    },
    pastel: {
      darkBlue: '#ABBBD5',
      lightOrange: '#FED2A9',
      lightGreen: '#CEE7A5',
      lightPurple: '#B3B5E9',
      lightRed: '#FF9D88',
      darkGreen: '#95CBAA',
      pink: '#FF82A0'
    },
    neutral: {
      black: '#111820',
      darkGrey: '#696969',
      normalGrey: '#a0a0a0',
      grey: '#ccc',
      lightGrey: '#efefef',
      extraLightGrey: '#f8f8f8',
      white: '#fff'
    },
    shades: {
      primary: {
        orange: (opacity: number) => `rgba(245, 128, 49, ${opacity})`
      },
      secondary: {
        green: (opacity: number) => `rgba(0, 128, 19, ${opacity})`,
        lightGreen: (opacity: number) => `rgba(147, 211, 118, ${opacity})`,
        red: (opacity: number) => `rgba(231, 77, 60, ${opacity})`,
        darkBlue: (opacity: number) => `rgba(29, 34, 54, ${opacity})`,
        blue: (opacity: number) => `rgba(52, 152, 219, ${opacity})`
      },
      pastel: {
        lightOrange: (opacity: number) => `rgba(254, 210, 169, ${opacity})`,
        lightGreen: (opacity: number) => `rgba(206, 231,165, ${opacity})`,
        lightRed: (opacity: number) => `rgba(255,158,136, ${opacity})`,
        darkGreen: (opacity: number) => `rgba(149,203,170, ${opacity})`,
        pink: (opacity: number) => `rgba(255, 13, 159, ${opacity})`
      },
      neutral: {
        black: (opacity: number) => `rgba(0, 0, 0, ${opacity})`,
        darkGrey: (opacity: number) => `rgba(105, 105, 105, ${opacity})`,
        grey: (opacity: number) => `rgba(204, 204, 204, ${opacity})`,
        lightGrey: (opacity: number) => `rgba(239, 239, 239, ${opacity})`,
        extraLightGrey: (opacity: number) => `rgba(248, 248, 248, ${opacity})`,
        white: (opacity: number) => `rgba(255, 255, 255, ${opacity})`
      }
    },
    linear: {
      darkBlue: `linear-gradient(89.66deg, #1D2236 0.42%, rgba(29, 34, 54, 0.8) 32.52%, rgba(29, 34, 54, 0.4) 99.84%)`,
      black: `linear-gradient(180deg, rgba(17, 24, 32, 0.4) 0%, #111820 100%)`,
      lightOrange: `linear-gradient(rgba(245, 143, 49, 0.71), rgba(245, 167, 49, 0.18))`,
      extraLightGrey: `linear-gradient(0deg, #f8f8f8, #f8f8f8) #000000`
    }
  },
  effects: {
    dropShadow: {
      elev1: '0 0.2rem 0.4rem rgba(0, 0, 0, 0.04)',
      elev2: '0 0.4rem 0.8rem rgba(0, 0, 0, 0.04)',
      elev3: '0 0.8rem 1.6rem rgba(0, 0, 0, 0.08)',
      elev4: '0 0.2rem 0.4rem 0 rgba(0, 0, 0, 0.25)',
      leftSidebars: '0.4rem 0 0.8rem rgba(0, 0, 0, 0.04)',
      rightSidebars: '-0.4rem 0 0.8rem rgba(0, 0, 0, 0.04)'
    }
  },
  layers: {
    dragIcon: 2,
    toast: 2,
    testerMessage: 1,
    tooltip: 1,
    popperDialog: 2,
    popperBackdrop: 2,
    sidePanel: 2,
    spinner: 1,
    headerMultiCompaniesModal: 1,
    headerMultiCompaniesPopper: 3,
    appBar: 1,
    backdrop: 1,
    profilePageHeader: 1,
    profilePageCard: 1
  },
  spacings: {
    xxxsmall: '0.4rem',
    xxsmall: '0.8rem',
    xsmall: '1.6rem',
    small: '2.4rem',
    medium: '3.2rem',
    large: '4.0rem',
    xlarge: '4.8rem',
    xxlarge: '5.6rem'
  },
  transition: {
    default: '0.3s ease-in-out',
    fast: '0.1s ease-in-out',
    slow: '1s ease-in-out'
  }
};
