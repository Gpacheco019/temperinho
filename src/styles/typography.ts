import { css } from 'styled-components';

export const typography = {
  SmallText: {
    Caption1: css`
      font-size: ${({ theme }) => theme.font.sizes.xsmall};
      font-weight: ${({ theme }) => theme.font.regular};
      line-height: ${({ theme }) => theme.spacings.xsmall};
    `,
    Caption2: css`
      font-size: 1rem;
      line-height: ${({ theme }) => theme.spacings.xsmall};
      font-weight: ${({ theme }) => theme.font.regular};
    `,
    Link1: css`
      font-size: ${({ theme }) => theme.font.sizes.xsmall};
      font-weight: ${({ theme }) => theme.font.medium};
      line-height: ${({ theme }) => theme.spacings.xsmall};
    `
  },
  Subtitles: {
    Sub1: css`
      display: block;
      font-size: ${({ theme }) => theme.font.sizes.medium};
      font-weight: ${({ theme }) => theme.font.medium};
      line-height: 2.4rem;
    `,
    Sub2: css`
      font-size: ${({ theme }) => theme.font.sizes.small};
      font-weight: ${({ theme }) => theme.font.medium};
      line-height: ${({ theme }) => theme.spacings.small};
    `
  },
  Text: {
    Body1: css`
      font-size: ${({ theme }) => theme.font.sizes.medium};
      line-height: ${({ theme }) => theme.spacings.small};
      font-weight: ${({ theme }) => theme.font.regular};
    `,
    Body2: css`
      font-size: ${({ theme }) => theme.font.sizes.small};
      line-height: 2rem;
      font-weight: ${({ theme }) => theme.font.regular};
    `,
    Body2Semibold: css`
      font-size: ${({ theme }) => theme.font.sizes.small};
      line-height: 2rem;
      font-weight: ${({ theme }) => theme.font.semibold};
    `,
    Link2: css`
      font-size: ${({ theme }) => theme.font.sizes.small};
      line-height: ${({ theme }) => theme.spacings.xsmall};
      font-weight: ${({ theme }) => theme.font.regular};
    `
  },
  Titles: {
    H1: css`
      font-weight: ${({ theme }) => theme.font.medium};
      font-size: 4.8rem;
      line-height: ${({ theme }) => theme.spacings.xxlarge};
    `,
    H2: css`
      font-weight: ${({ theme }) => theme.font.medium};
      font-size: 4rem;
      line-height: ${({ theme }) => theme.spacings.xlarge};
    `,
    H3: css`
      font-weight: ${({ theme }) => theme.font.medium};
      font-size: 3.2rem;
      line-height: ${({ theme }) => theme.spacings.large};
    `,
    H4: css`
      font-weight: ${({ theme }) => theme.font.medium};
      font-size: 2.4rem;
      line-height: ${({ theme }) => theme.spacings.medium};
    `,
    H5: css`
      font-weight: ${({ theme }) => theme.font.medium};
      font-size: ${({ theme }) => theme.font.sizes.xlarge};
      line-height: ${({ theme }) => theme.spacings.small};
    `,
    H6: css`
      font-size: ${({ theme }) => theme.font.sizes.large};
      font-weight: ${({ theme }) => theme.font.medium};
      line-height: ${({ theme }) => theme.spacings.small};
    `
  },
  Buttons: {
    Button1: css`
      text-decoration: none;
      background-color: transparent;
      text-transform: uppercase;
      letter-spacing: 0.1rem;
      border: none;
      font-size: ${({ theme }) => theme.font.sizes.small};
      font-weight: ${({ theme }) => theme.font.medium};
      line-height: ${({ theme }) => theme.spacings.small};
      cursor: pointer;
    `,
    Button2: css`
      text-decoration: none;
      background-color: transparent;
      text-transform: uppercase;
      letter-spacing: 0.1rem;
      border: none;
      font-size: ${({ theme }) => theme.font.sizes.xsmall};
      font-weight: ${({ theme }) => theme.font.medium};
      line-height: ${({ theme }) => theme.spacings.small};
      cursor: pointer;
    `
  }
};
