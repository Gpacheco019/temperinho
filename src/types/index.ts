export * from './colorVariation';
export * from './fields';
export * from './files';
