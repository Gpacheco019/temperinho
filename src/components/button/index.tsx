import React, { MouseEvent, forwardRef, useEffect, useState } from 'react';

import * as S from './styles';
import { ButtonProps } from './types';

const ButtonComponent = (
  {
    customLayoutColor,
    children,
    icon,
    size = 'small',
    fullWidth = false,
    square = false,
    disabled = false,
    loading = false,
    circle = false,
    variant = 'primary',
    colorVariation,
    iconPosition = 'right',
    timeout,
    onClick,
    ...props
  }: ButtonProps,
  ref: React.Ref<HTMLButtonElement>
) => {
  const [waitingTime, setWaitingTime] = useState(timeout);
  const timer: NodeJS.Timer = setInterval(() => {
    setWaitingTime(waitingTime! - 1);
  }, 1000);

  const handleClick = (event?: MouseEvent<HTMLButtonElement>) => {
    typeof timeout !== 'undefined' && setWaitingTime(timeout);
    onClick && onClick(event);
  };

  useEffect(() => {
    if (waitingTime! < 1) {
      clearInterval(timer);
    }
  }, [waitingTime, timer]);

  useEffect(() => {
    return () => {
      clearInterval(timer);
    };
  });

  return (
    <S.Wrapper
      customLayoutColor={customLayoutColor}
      size={size}
      fullWidth={fullWidth}
      hasIcon={!!icon}
      square={square}
      ref={ref}
      disabled={disabled || loading || !!waitingTime}
      variant={variant}
      colorVariation={colorVariation}
      circle={circle}
      onClick={handleClick}
      {...props}
    >
      {!!icon && <S.Icon iconPosition={iconPosition}>{icon}</S.Icon>}
      {!loading && !waitingTime && !!children && <>{children}</>}
      {!!waitingTime && (
        <>
          {children}
          {waitingTime && ` (${waitingTime})`}
        </>
      )}
      {!!loading && <S.Loading aria-hidden={loading} aria-label="carregando" />}
    </S.Wrapper>
  );
};
export const Button = forwardRef<HTMLButtonElement, ButtonProps>(
  ButtonComponent
);
