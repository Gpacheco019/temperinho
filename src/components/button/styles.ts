import { motion } from 'framer-motion';
import styled, { css, DefaultTheme } from 'styled-components';

import { getColorByVariant } from '../../utils';

import { typography } from '../../styles';
import { ButtonProps, WrapperProps } from './types';

const spinTransition = {
  loop: Infinity,
  ease: 'linear',
  duration: 1
};

const wrapperModifiers = {
  mini: (theme: DefaultTheme) => css`
    height: ${theme.spacings.small};
    padding: ${theme.spacings.xxsmall} ${theme.spacings.xsmall};
  `,
  small: (theme: DefaultTheme) => css`
    height: ${theme.spacings.medium};
    padding: ${theme.spacings.xxsmall} ${theme.spacings.xsmall};
  `,
  big: (theme: DefaultTheme) => css`
    height: ${theme.spacings.large};
    padding: ${theme.spacings.xxsmall} ${theme.spacings.xsmall};
  `,
  primary: (
    theme: DefaultTheme,
    colorVariation: WrapperProps['colorVariation'],
    customLayoutColor: WrapperProps['customLayoutColor']
  ) => css`
    background-color: ${colorVariation
      ? getColorByVariant(theme, colorVariation)
      : customLayoutColor || theme.colors.primary.orange};
    color: ${theme.colors.neutral.white};
    border: 0;
  `,
  secondary: (
    theme: DefaultTheme,
    colorVariation: WrapperProps['colorVariation'],
    customLayoutColor: WrapperProps['customLayoutColor']
  ) => css`
    background-color: transparent;
    border: 0;
    color: ${colorVariation
      ? getColorByVariant(theme, colorVariation)
      : customLayoutColor || theme.colors.primary.orange};
  `,
  third: (
    theme: DefaultTheme,
    colorVariation: WrapperProps['colorVariation'],
    customLayoutColor: WrapperProps['customLayoutColor']
  ) => css`
    background-color: transparent;
    color: ${colorVariation
      ? getColorByVariant(theme, colorVariation)
      : customLayoutColor || theme.colors.primary.orange};
    border: 0.1rem solid
      ${colorVariation
        ? getColorByVariant(theme, colorVariation)
        : customLayoutColor || theme.colors.primary.orange};
  `,
  withIcon: (theme: DefaultTheme) => css`
    svg {
      & + span {
        margin-left: ${theme.spacings.xxxsmall};
      }
    }
  `,
  fullWidth: () => css`
    width: 100%;
  `,
  square: (theme: DefaultTheme) => css`
    width: 3.6rem;
    height: 3.6rem;
    padding: ${theme.spacings.xxsmall};

    &:hover {
      filter: brightness(90%);
      transition: filter 0.2s;
    }
  `,
  circle: (theme: DefaultTheme) => css`
    border-radius: 50%;
    padding: ${theme.spacings.xxsmall};
  `,
  disabled: () => css`
    &:disabled {
      cursor: not-allowed;
      opacity: 0.3;
    }
  `
};

export const Wrapper = styled.button<WrapperProps>`
  ${({
    theme,
    customLayoutColor,
    size,
    fullWidth,
    hasIcon,
    square,
    disabled,
    variant,
    colorVariation,
    circle
  }) => css`
    ${typography.Buttons.Button1}

    ${!!size && wrapperModifiers[size](theme)};
    ${!!variant &&
    wrapperModifiers[variant](theme, colorVariation, customLayoutColor)};
    ${!!hasIcon && wrapperModifiers.withIcon(theme)};
    ${circle
      ? wrapperModifiers.circle(theme)
      : `border-radius: ${theme.spacings.xxxsmall};`};
    ${!!fullWidth && wrapperModifiers.fullWidth()};
    ${!!square && wrapperModifiers.square(theme)};
    ${disabled && wrapperModifiers.disabled()};
  `}

  display: inline-flex;
  align-items: center;
  justify-content: center;
  cursor: pointer;
`;

export const Icon = styled.div<ButtonProps>`
  display: flex;
  order: ${({ iconPosition }) => (iconPosition === 'right' ? 1 : 0)};
`;

export const Loading = styled(motion.span).attrs({
  transition: spinTransition,
  animate: { rotate: 360 }
})`
  ${({ theme }) => css`
    display: block;
    width: 2rem;
    height: 2rem;
    border: 0.2rem solid ${theme.colors.neutral.darkGrey};
    border-top: 0.2rem solid ${theme.colors.neutral.white};
    border-radius: 50%;
    box-sizing: border-box;
  `}
`;
