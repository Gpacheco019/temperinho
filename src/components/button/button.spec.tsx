import React from 'react';
import { MdAdd } from 'react-icons/md';

import { act, fireEvent } from '@testing-library/react';

import { createSetupToTests } from '../../utils/';

import { Button } from '../../components/button';
import { ButtonProps } from '../../components/button/types';
import { theme } from '../../styles/index';

const setup = createSetupToTests<ButtonProps>({ children: 'Click me' }, Button);

describe('<Button />', () => {
  it('should render the small size by default', () => {
    const { container, getByRole } = setup();

    expect(container.firstChild).toMatchSnapshot();
    expect(getByRole('button', { name: /Click me/i })).toHaveStyle({
      height: theme.spacings.medium,
      padding: `${theme.spacings.xxsmall} ${theme.spacings.xsmall}`
    });
  });

  it('should render the mini size', () => {
    const { getByRole } = setup({ size: 'mini' });

    expect(getByRole('button', { name: /Click me/i })).toHaveStyle({
      height: theme.spacings.small
    });
  });

  it('should render the big size', () => {
    const { getByRole } = setup({ size: 'big' });

    expect(getByRole('button', { name: /Click me/i })).toHaveStyle({
      height: theme.spacings.large,
      padding: `${theme.spacings.xxsmall} ${theme.spacings.xsmall}`
    });
  });

  it('should render primary variant', () => {
    const { getByRole } = setup({ square: true });

    expect(getByRole('button', { name: /Click me/i })).toHaveStyle({
      backgroundColor: theme.colors.primary.orange,
      color: theme.colors.neutral.white
    });
  });

  it('should render secondary variant', () => {
    const { getByRole } = setup({
      variant: 'secondary',
      colorVariation: 'neutral.darkGrey'
    });

    expect(getByRole('button', { name: /Click me/i })).toHaveStyle({
      color: theme.colors.neutral.darkGrey
    });
  });

  it('should render third button variant', () => {
    const { getByRole } = setup({
      variant: 'third',
      colorVariation: 'neutral.darkGrey'
    });

    expect(getByRole('button', { name: /Click me/i })).toHaveStyle({
      color: theme.colors.neutral.darkGrey,
      border: `0.1rem solid ${theme.colors.neutral.darkGrey}`
    });
  });

  it('should render a fullWidth variant', () => {
    const { getByRole } = setup({ fullWidth: true });

    expect(getByRole('button', { name: /Click me/i })).toHaveStyle({
      width: '100%'
    });
  });

  it('should render an icon variant', () => {
    const { getByText, getByTestId } = setup({
      icon: <MdAdd data-testid="icon" />
    });

    expect(getByText(/Click me/i)).toBeInTheDocument();
    expect(getByTestId('icon')).toBeInTheDocument();
  });

  it('should render a square variant', () => {
    const { getByRole } = setup({ square: true });

    expect(getByRole('button', { name: /Click me/i })).toHaveStyle({
      backgroundColor: theme.colors.primary.orange
    });
  });

  it('should render a circle variant', () => {
    const { getByRole } = setup({ square: true, circle: true });

    expect(getByRole('button', { name: /Click me/i })).toHaveStyle({
      borderRadius: '50%'
    });
  });

  it('should render a disabled Button', () => {
    const { getByRole } = setup({ disabled: true });

    expect(getByRole('button', { name: /Click me/i })).toHaveStyleRule(
      'cursor',
      'not-allowed',
      {
        modifier: ':disabled'
      }
    );
  });

  it('should render a loading Button', () => {
    const { getByLabelText } = setup({ disabled: true, loading: true });

    expect(getByLabelText('carregando')).toBeInTheDocument();
  });

  describe(' Button with timeout', () => {
    it('should render a Button with timeout', () => {
      const { getByText } = setup({ timeout: 10, children: 'texto' });

      expect(getByText('texto (10)')).toBeInTheDocument();
    });

    it('should render a Button with timeout', () => {
      const onClick = jest.fn();
      const { getByText } = setup({
        timeout: 0,
        children: 'texto',
        onClick: onClick
      });
      const textButton = getByText('texto');

      expect(textButton).toBeInTheDocument();

      fireEvent.click(textButton);
      expect(onClick).toBeCalled();
      expect(onClick).toBeCalledWith(expect.anything());
    });

    it('should render with timer running', () => {
      jest.useFakeTimers();
      jest.spyOn(global, 'setInterval');
      const handleSubmit = jest.fn();
      setup({
        onClick: handleSubmit,
        timeout: 1
      });
      act(() => {
        jest.runOnlyPendingTimers();
      });

      expect(setInterval).toHaveBeenCalledTimes(2);
    });
  });
});
