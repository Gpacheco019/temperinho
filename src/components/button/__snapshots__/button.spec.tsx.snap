// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Button /> should render the small size by default 1`] = `
.c0 {
  -webkit-text-decoration: none;
  text-decoration: none;
  background-color: transparent;
  text-transform: uppercase;
  -webkit-letter-spacing: 0.1rem;
  -moz-letter-spacing: 0.1rem;
  -ms-letter-spacing: 0.1rem;
  letter-spacing: 0.1rem;
  border: none;
  font-size: 1.4rem;
  font-weight: 500;
  line-height: 2.4rem;
  cursor: pointer;
  height: 3.2rem;
  padding: 0.8rem 1.6rem;
  background-color: #f58031;
  color: #fff;
  border: 0;
  border-radius: 0.4rem;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  cursor: pointer;
}

<button
  class="c0"
>
  Click me
</button>
`;
